File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionRotateTo.java
package br.com.insanitech.spritekit;

/**
 * Created by anderson on 06/01/17.
 */
class SKActionRotateTo extends SKAction {
    private float startRadians;
    private float radians;

    public SKActionRotateTo(float radians) {
        this.radians = radians;
    }

    @Override
    void computeStart() {
        startRadians = getParent().zRotation;
    }

    @Override
    void computeAction(long elapsed) {
        switch (timingMode) {
            case EaseIn: {
                getParent().zRotation = SKEaseCalculations.easeIn(elapsed, startRadians, radians - startRadians, duration);
            }
            break;

            case EaseOut: {
                getParent().zRotation = SKEaseCalculations.easeIn(elapsed, startRadians, radians - startRadians, duration);
            }
            break;

            case EaseInEaseOut: {
                getParent().zRotation = SKEaseCalculations.easeIn(elapsed, startRadians, radians - startRadians, duration);
            }
            break;

            case Linear: {
                getParent().zRotation = SKEaseCalculations.easeIn(elapsed, startRadians, radians - startRadians, duration);
            }
            break;
        }
    }

    @Override
    void computeFinish() {
        getParent().zRotation = radians;
    }

    @Override
    boolean willHandleFinish() {
        return false;
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionRotateTo.kt
package br.com.insanitech.spritekit

/**
 * Created by anderson on 06/01/17.
 */
internal class SKActionRotateTo(private val radians: Float) : SKAction() {
    private var startRadians: Float = 0.toFloat()

    internal override fun computeStart() {
        startRadians = parent?.zRotation ?: 0.0f
    }

    internal override fun computeAction(elapsed: Long) {
        when (timingMode) {
            SKActionTimingMode.EaseIn -> {
                parent?.zRotation = SKEaseCalculations.easeIn(elapsed.toFloat(), startRadians, radians - startRadians, duration.toFloat())
            }

            SKActionTimingMode.EaseOut -> {
                parent?.zRotation = SKEaseCalculations.easeIn(elapsed.toFloat(), startRadians, radians - startRadians, duration.toFloat())
            }

            SKActionTimingMode.EaseInEaseOut -> {
                parent?.zRotation = SKEaseCalculations.easeIn(elapsed.toFloat(), startRadians, radians - startRadians, duration.toFloat())
            }

            SKActionTimingMode.Linear -> {
                parent?.zRotation = SKEaseCalculations.easeIn(elapsed.toFloat(), startRadians, radians - startRadians, duration.toFloat())
            }
        }
    }

    internal override fun computeFinish() {
        parent?.zRotation = radians
    }
}
